compiler/rustc_resolve/src/access_levels.rs

```diff
@@ -72,33 +70,60 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
 
-        if let Some(exports) = self.r.reexport_map.get(&module_id) {
-            let pub_exports = exports
-                .iter()
-                .filter(|ex| ex.vis == Visibility::Public)
-                .cloned()
-                .collect::<Vec<_>>();
+        for (.., name_resolution) in resolutions.borrow_mut().iter() {
+            if let Some(binding) = name_resolution.borrow().binding() {
+                if let Some(res) = self.r.is_reexport(binding) && binding.vis.is_public() {
+                    let module_level = self.r.access_levels.map.get(&module_id).copied();
+                    assert!(module_level >= Some(AccessLevel::Exported));
 
-            let module = self.r.get_module(module_id.to_def_id()).unwrap();
-            for export in pub_exports.into_iter() {
-                if let Some(export_def_id) = export.res.opt_def_id().and_then(|id| id.as_local()) {
-                    self.set_access_level_def_id(export_def_id, Some(AccessLevel::Exported));
+                    if let Some(def_id) = res.opt_def_id().and_then(|id| id.as_local()) {
+                        self.set_access_level_def_id(def_id, Some(AccessLevel::Exported));
+                        set_import_binding_access_level(self, binding, module_level);
+                    }
+                    continue;
                 }
 
-                if let Some(ns) = export.res.ns() {
-                    let key = BindingKey { ident: export.ident, ns, disambiguator: 0 };
-                    let name_res = self.r.resolution(module, key);
-                    if let Some(binding) = name_res.borrow().binding() {
-                        set_import_binding_access_level(self, binding, module_level)
+                let res = binding.res();
+                if let Some(def_id) = res.opt_def_id().and_then(|id| id.as_local()) {
+                    match binding.res() {
+                        Res::Def(def_kind, _) => match def_kind {
+                            DefKind::ForeignTy
+                            | DefKind::ExternCrate
+                            | DefKind::Static(..)
+                            | DefKind::Const
+                            | DefKind::Fn
+                            | DefKind::TyAlias
+                            | DefKind::Enum
+                            | DefKind::Struct
+                            | DefKind::Union
+                            | DefKind::Trait
+                            | DefKind::TraitAlias
+                            | DefKind::AssocFn
+                            | DefKind::AssocConst
+                            | DefKind::AssocTy
+                            | DefKind::Macro(..) => {
+                                if binding.vis.is_public() {
+                                    self.set_access_level_def_id(def_id, self.prev_level);
+                                } else {
+                                    self.set_access_level_def_id(def_id, None);
```

Comment:
`set_access_level_def_id` cannot decrease the level, so this line is always noop.